modules/social_features/social_profile/src/Service/SocialProfileAccessService.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+namespace Drupal\social_profile\Service;
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Access\AccessResultInterface;
+use Drupal\Core\DependencyInjection\ContainerInjectionInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\user\UserInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provide a service for Profile Access service.
+ *
+ * @package Drupal\social_profile\Service
+ */
+class SocialProfileAccessService implements ContainerInjectionInterface {
+
+  /**
+   * The current route match.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected RouteMatchInterface $routeMatch;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected EntityTypeManagerInterface $entityTypeManager;
+
+  /**
+   * SocialProfileAccessService constructor.
+   *
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route match.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   */
+  public function __construct(RouteMatchInterface $route_match, EntityTypeManagerInterface $entity_type_manager) {
+    $this->routeMatch = $route_match;
+    $this->entityTypeManager = $entity_type_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container): self {
+    return new static(
+      $container->get('current_route_match'),
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * Validation permission for profile pages.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   Logged user.
+   *
+   * @return \Drupal\Core\Access\AccessResultInterface
+   *   Return TRUE when user has access.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function access(AccountInterface $account): AccessResultInterface {
+    // Load the user entity.
+    $user_id = $this->routeMatch->getRawParameter('user');
+    $user = $this->entityTypeManager->getStorage('user')
+      ->load($user_id);
+    if (!$user instanceof UserInterface) {
+      return AccessResult::forbidden();
+    }
+
+    // If user is blocked, check special permission.
+    if ($user->isBlocked()) {
+      return AccessResult::allowedIfHasPermissions($account, ['view blocked user']);
+    }
+
+    // Check if current page is from logged user.
+    if ($account->id() === $user->id()) {
+      return AccessResult::allowedIfHasPermissions($account, ['view own profile profile']);
+    }
+
+    // Check if logged user can view all profiles.
+    return AccessResult::allowedIfHasPermissions($account, ['view any profile profile']);
```

Comment:
How does this logic differ from what would be evaluated if we called `$profile->access('view', $account);`?

What happens if other modules change when a user can be viewed by using Drupal's access APIs?

Reply:
I debugged with entity-access instead of custom-access and I found where and why the isn't give access.

The validation happen here: html/core/lib/Drupal/Core/Entity/EntityAccessCheck.php:65
The code try to found a parameter with profile, but the view has user as parameter, so to work we need to add this on view.

![image](https://github.com/user-attachments/assets/7e7bcdd7-c0f6-4173-8d4b-ac808a3049d1)